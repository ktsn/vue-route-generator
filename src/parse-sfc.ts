import type { parseComponent as _parserV2 } from 'vue-template-compiler'
import type { parse as _parserV3 } from '@vue/compiler-sfc'

export interface ParseResult {
  customBlocks: CustomBlock[]
}

export interface CustomBlock {
  type: string
  content: string
}

export function parseSFC(code: string): ParseResult {
  try {
    const parserV2: typeof _parserV2 = require('vue-template-compiler')
      .parseComponent
    return parserV2(code, {
      pad: 'space',
    })
  } catch {
    try {
      const parserV3: typeof _parserV3 = require('@vue/compiler-sfc').parse
      return parserV3(code, {
        pad: 'space',
      }).descriptor
    } catch {
      throw new Error(
        '[vue-route-generator] Either "vue-template-compiler" or "@vue/compiler-sfc" is required.'
      )
    }
  }
}
