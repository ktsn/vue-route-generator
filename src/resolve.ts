import { NestedMap, setToMap } from './nested-map'
import { parseSFC } from './parse-sfc'

export interface PageMeta {
  name: string
  chunkName: string
  specifier: string
  path: string
  pathSegments: string[]
  component: string
  children?: PageMeta[]
  routeMeta?: any
  route?: any
}

interface FileError extends Error {
  file?: string
}

const routeMetaName = 'route-meta'
const routeBlockName = 'route'

export function resolveRoutePaths(
  paths: string[],
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string
): PageMeta[] {
  const map: NestedMap<string[]> = {}

  const splitted = paths.map((p) => p.split('/'))
  splitted.forEach((path) => {
    setToMap(map, pathToMapPath(path), path)
  })

  return pathMapToMeta(map, importPrefix, nested, readFile)
}

function pathMapToMeta(
  map: NestedMap<string[]>,
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string,
  parentDepth: number = 0
): PageMeta[] {
  if (map.value) {
    const path = map.value

    const meta: PageMeta = {
      name: pathToName(path),
      chunkName: pathToChunkName(path),
      specifier: pathToSpecifier(path),
      path: pathToRoute(path, parentDepth, nested),
      pathSegments: toActualPath(path),
      component: importPrefix + path.join('/'),
    }

    const content = readFile(path.join('/'))
    const parsed = parseSFC(content)
    const routeMetaBlock = parsed.customBlocks.find(
      (b) => b.type === routeMetaName
    )
    const routeBlock = parsed.customBlocks.find(
      (b) => b.type === routeBlockName
    )

    // Deprecated. Will be removed in a later version
    if (routeMetaBlock) {
      console.warn(
        '<route-meta> custom block is deprecated. Use <route> block instead. Found in ' +
          path.join('/')
      )
      meta.routeMeta = tryParseCustomBlock(
        routeMetaBlock.content,
        path,
        'route-meta'
      )
    }

    if (routeBlock) {
      meta.route = tryParseCustomBlock(routeBlock.content, path, 'route')
    }

    if (map.children) {
      meta.children = pathMapChildrenToMeta(
        map.children,
        importPrefix,
        nested,
        readFile,
        meta.pathSegments.length
      )
    }

    return [meta]
  }

  return map.children
    ? pathMapChildrenToMeta(
        map.children,
        importPrefix,
        nested,
        readFile,
        parentDepth
      )
    : []
}

function routePathComparator(a: string[], b: string[]): number {
  const a0 = a[0]
  const b0 = b[0]

  if (!a0 || !b0) {
    return a.length - b.length
  }

  const aOrder = isDynamicRoute(a0) ? 1 : 0
  const bOrder = isDynamicRoute(b0) ? 1 : 0
  const order = aOrder - bOrder

  return order !== 0 ? order : routePathComparator(a.slice(1), b.slice(1))
}

function pathMapChildrenToMeta(
  children: Map<string, NestedMap<string[]>>,
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string,
  parentDepth: number
): PageMeta[] {
  return Array.from(children.values())
    .reduce<PageMeta[]>((acc, value) => {
      return acc.concat(
        pathMapToMeta(value, importPrefix, nested, readFile, parentDepth)
      )
    }, [])
    .sort((a, b) => {
      // Prioritize static routes than dynamic routes
      return routePathComparator(a.pathSegments, b.pathSegments)
    })
}

function tryParseCustomBlock(
  content: string,
  filePath: string[],
  blockName: string
): any {
  try {
    return JSON.parse(content)
  } catch (err) {
    const joinedPath = filePath.join('/')
    const wrapped: FileError = new Error(
      `Invalid json format of <${blockName}> content in ${joinedPath}\n` +
        err.message
    )

    // Store file path to provide useful information to downstream tools
    // like friendly-errors-webpack-plugin
    wrapped.file = joinedPath

    throw wrapped
  }
}

function isDynamicRoute(segment: string): boolean {
  return segment[0] === ':'
}

function isOmittable(segment: string): boolean {
  return segment === 'index'
}

/**
 * - Remove `.vue` from the last path
 * - Omit if the last segument is `index`
 * - Convert dynamic route to `:param` format
 */
function toActualPath(segments: string[]): string[] {
  const lastIndex = segments.length - 1
  const last = basename(segments[lastIndex])
  segments = segments.slice(0, -1).concat(last)

  return segments
    .map((s, i) => {
      if (s[0] === '_') {
        const suffix = lastIndex === i ? '?' : ''
        return ':' + s.slice(1) + suffix
      } else {
        return s
      }
    })
    .filter((s) => !isOmittable(s))
}

function pathToMapPath(segments: string[]): string[] {
  const last = segments[segments.length - 1]
  return segments.slice(0, -1).concat(basename(last))
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments
    .slice(0, -1)
    .concat(basename(last))
    .filter((s) => !isOmittable(s))

  if (segments.length === 0) {
    return 'index'
  }

  return segments
    .map((s) => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToChunkName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments.slice(0, -1).concat(basename(last))

  return segments
    .map((s) => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const last = segments[segments.length - 1]
  const replaced = segments
    .slice(0, -1)
    .concat(basename(last))
    .join('_')
    .replace(/[^a-zA-Z0-9]/g, '_')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(
  segments: string[],
  parentDepth: number,
  nested: boolean
): string {
  const prefix = nested || parentDepth > 0 ? '' : '/'
  return prefix + toActualPath(segments).slice(parentDepth).join('/')
}

function basename(filename: string): string {
  return filename.replace(/\.[^.]+$/g, '')
}
