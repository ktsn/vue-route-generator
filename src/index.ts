import * as fs from 'fs'
import * as path from 'path'
import * as fg from 'fast-glob'
import { createRoutes } from './template/routes'
import { resolveRoutePaths } from './resolve'

export interface GenerateConfig {
  pages: string
  importPrefix?: string
  dynamicImport?: boolean
  chunkNamePrefix?: string
  nested?: boolean
  inlineRouteBlock?: boolean
}

export function generateRoutes({
  pages,
  importPrefix = '@/pages/',
  dynamicImport = true,
  chunkNamePrefix = '',
  nested = false,
  inlineRouteBlock = true,
}: GenerateConfig): string {
  const patterns = ['**/*.vue', '!**/__*__.vue', '!**/__*__/**']

  const pagePaths = fg.sync(patterns, {
    cwd: pages,
    onlyFiles: true,
  })

  const metaList = resolveRoutePaths(pagePaths, {
    importPrefix,
    nested,
    inlineRouteBlock,
    readFile: (file) => {
      return fs.readFileSync(path.join(pages, file), 'utf8')
    },
  })

  return createRoutes(
    metaList,
    dynamicImport,
    chunkNamePrefix,
    inlineRouteBlock
  )
}
