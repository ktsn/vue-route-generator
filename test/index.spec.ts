import * as path from 'path'
import { generateRoutes } from '../src/index'

it('Integration test', () => {
  const res = generateRoutes({
    pages: path.resolve(__dirname, 'fixtures'),
  })

  expect(res).toMatchSnapshot()
})
