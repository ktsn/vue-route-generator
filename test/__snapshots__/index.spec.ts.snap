// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test 1`] = `
"function index() {
  return import(
  /* webpackChunkName: \\"index\\" */
  '@/pages/index.vue');
}

function foo() {
  return import(
  /* webpackChunkName: \\"foo\\" */
  '@/pages/foo.vue');
}

function foo_index() {
  return import(
  /* webpackChunkName: \\"foo-index\\" */
  '@/pages/foo/index.vue');
}

function foo_bar() {
  return import(
  /* webpackChunkName: \\"foo-bar\\" */
  '@/pages/foo/bar.vue');
}

function foo__id() {
  return import(
  /* webpackChunkName: \\"foo-id\\" */
  '@/pages/foo/_id.vue');
}

function route() {
  return import(
  /* webpackChunkName: \\"route\\" */
  '@/pages/route.vue');
}

function user_register() {
  return import(
  /* webpackChunkName: \\"user.register\\" */
  '@/pages/user.register.vue');
}

function _id_index() {
  return import(
  /* webpackChunkName: \\"id-index\\" */
  '@/pages/_id/index.vue');
}

function _id_test() {
  return import(
  /* webpackChunkName: \\"id-test\\" */
  '@/pages/_id/test.vue');
}

export default [{
  name: 'index',
  path: '/',
  component: index
}, {
  path: '/foo',
  component: foo,
  children: [{
    name: 'foo',
    path: '',
    component: foo_index
  }, {
    name: 'foo-bar',
    path: 'bar',
    component: foo_bar
  }, {
    name: 'foo-id',
    path: ':id?',
    component: foo__id
  }]
}, {
  name: 'Test',
  path: '/route',
  component: route,
  meta: {
    \\"requiresAuth\\": true
  }
}, {
  name: 'user.register',
  path: '/user.register',
  component: user_register
}, {
  name: 'id',
  path: '/:id',
  component: _id_index
}, {
  name: 'id-test',
  path: '/:id/test',
  component: _id_test
}];"
`;
